PubnativeLite/PubnativeLiteDemo/Adapters/AppMonet Adapters/MoPub Adapters/AppMonet+MoPub.m

```diff
@@ -23,45 +23,73 @@
-
-@property (class, nonatomic) NSMutableDictionary *dict;
-@property (class, nonatomic) NSMutableDictionary *interstitialDict;
-@property (class, nonatomic, strong) HyBidAdRequest *adRequest;
-@property (class, nonatomic, strong) HyBidInterstitialAdRequest *interstitialAdRequest;
+typedef struct { MPAdView *adView;} AppMonetAdStruct;
+typedef struct { MPInterstitialAdController *interstitial;} AppMonetInterstitialAdStruct;
 
 @end
 
 @implementation AppMonet (MoPub)
 
+static void (^_onReadyBlock)(void);
+
+static NSMutableDictionary *_dict=nil;
++ (NSMutableDictionary *)dict {
+  if (_dict == nil) {
+    _dict = [[NSMutableDictionary alloc] init];
+  }
+  return _dict;
+}
+
+static NSMutableDictionary *_interstitialDict=nil;
++ (NSMutableDictionary *)interstitialDict {
+  if (_interstitialDict == nil) {
+      _interstitialDict = [[NSMutableDictionary alloc] init];
+  }
+  return _interstitialDict;
+}
+
+static HyBidAdRequest *_adRequest=nil;
++ (HyBidAdRequest *)adRequest {
+  if (_adRequest == nil) {
+      _adRequest = [[HyBidAdRequest alloc] init];
+  }
+  return _adRequest;
+}
+
+static HyBidInterstitialAdRequest *_interstitialAdRequest=nil;
++ (HyBidInterstitialAdRequest *)interstitialAdRequest {
+  if (_interstitialAdRequest == nil) {
+      _interstitialAdRequest = [[HyBidInterstitialAdRequest alloc] init];
+  }
+  return _interstitialAdRequest;
+}
+
 + (void)addBids:(MPAdView *)adView andAppMonetAdUnitId:(NSString *)appMonetAdUinitId andTimeout:(NSNumber *)timeout :(void (^)(void))onReadyBlock
 {
     AppMonetAdStruct appMonetStruct;
     appMonetStruct.adView = adView;
-    appMonetStruct.onReadyBlock = onReadyBlock;
-    [AppMonet.dict setObject:
-                      [NSValue valueWithPointer:&appMonetStruct]
+    _onReadyBlock = onReadyBlock;
+//    appMonetStruct.onReadyBlock = nil;
+    NSData *data = [NSData dataWithBytes:&appMonetStruct length:sizeof(AppMonetAdStruct)];
+    [AppMonet.dict setObject:data
                       forKey:appMonetAdUinitId];
     
-    HyBidAdRequest *request = [[HyBidAdRequest alloc] init];
-    request.adSize = HyBidAdSize.SIZE_300x250;
-    [request requestAdWithDelegate:(id<HyBidAdRequestDelegate>)self withZoneID:appMonetAdUinitId];
+    self.adRequest.adSize = HyBidAdSize.SIZE_300x250;
+    [self.adRequest requestAdWithDelegate:AppMonet.self withZoneID:appMonetAdUinitId];
 }
 
 + (void)addInterstitialBids:(MPInterstitialAdController *)interstitial andAppMonetAdUnitId:(NSString *)appMonetAdUinitId andTimeout:(NSNumber *)timeout :(void (^)(void))onReadyBlock __attribute__((deprecated))
 {
     AppMonetInterstitialAdStruct appMonetInterstitialStruct;
     appMonetInterstitialStruct.interstitial = interstitial;
-    appMonetInterstitialStruct.onReadyBlock = onReadyBlock;
-    [AppMonet.interstitialDict setObject:
-     [NSValue valueWithPointer:&appMonetInterstitialStruct]
-                      forKey:appMonetAdUinitId];
+    _onReadyBlock = onReadyBlock;
```

Comment:
the same goes here

Reply:
Ah yes. We discussed this. It's important to have it in the struct otherwise there could be concurrency problems.

Reply:
@eros902002 exactly :)

Reply:
fixed. check my last commit.